net/netcheck/netcheck.go

```diff
@@ -1086,12 +1096,86 @@ func (c *Client) measureHTTPSLatency(ctx context.Context, reg *tailcfg.DERPRegio
 	return result.ServerProcessing, ip, nil
 }
 
+func (c *Client) measureAllICMPLatency(ctx context.Context, rs *reportState, dm *tailcfg.DERPMap) error {
+	var wg sync.WaitGroup
+	var need []*tailcfg.DERPRegion
+	for rid, reg := range dm.Regions {
+		if !rs.haveRegionLatency(rid) && regionHasDERPNode(reg) {
+			need = append(need, reg)
+		}
+	}
+	if len(need) == 0 {
+		return nil
+	}
+	wg.Add(len(need))
+	c.logf("netcheck: UDP is blocked, trying ICMP")
+
+	p, err := ping.New(ctx, c.logf)
+	if err != nil {
+		return err
+	}
+	defer p.Close()
+
+	for _, reg := range need {
+		go func(reg *tailcfg.DERPRegion) {
+			defer wg.Done()
+			if d, err := c.measureICMPLatency(ctx, reg, p); err != nil {
+				c.logf("[v1] netcheck: measuring ICMP latency of %v (%d): %v", reg.RegionCode, reg.RegionID, err)
+			} else {
+				c.logf("[v1] netcheck: ICMP latency of %v (%d): %v", reg.RegionCode, reg.RegionID, d)
+				rs.mu.Lock()
+				rs.report.RegionLatency[reg.RegionID] = d
```

Comment:
is this map non-nil already?

If unsure, you can use `mak.Set(&rs.report.RegionLatency, reg.RegionID, d)`